### PR TITLE
WIP redirect flow

### PR DESCRIPTION
**Pull request by bdorfman-stripe**
Also deleted a bunch of deprecated code

### CODE REVIEW

File: Stripe/PublicHeaders/STPAPIClient.h [hunk 1]
```diff
@@ -153,66 +153,5 @@ static NSString *const STPSDKVersion = @"9.3.0";
 
 @end
 
-#pragma mark - Deprecated Methods
```
Comment (bg-stripe) [inline Stripe/PublicHeaders/STPAPIClient.h]: can we pull out the removal of deprecated stuff to another PR?

File: Stripe/STPRedirectClient.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+@interface STPRedirectClient : NSObject
+@property (nonatomic, readonly) STPRedirectConfiguration *configuration;
+@property (nonatomic, nullable, readonly) STPSource *inProgressAuthSource;
+
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithCoder:(NSCoder *)aDecoder NS_UNAVAILABLE;
+- (instancetype)initWithConfiguration:(STPRedirectConfiguration *)configuration NS_EXTENSION_UNAVAILABLE("Redirect based sources are not available in extensions");
+
+/**
+ Starts a new redirect flow authorization for the given source.
```
Comment (bg-stripe) [inline Stripe/STPRedirectClient.h]: nit: should probably match formatting with other headers:
```
/**
 *
 */
```

File: Stripe/STPRedirectClient.m [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    _inProgressAuthSource = nil;
+    self.completion = nil;
+
+}
+
+- (BOOL)handleURLCallback:(NSURL *)url {
+    if (url == nil
+        || self.inProgressAuthSource == nil) {
+        [self cleanupAndCompleteRedirectWithError:nil];
```
Comment (bg-stripe) [inline Stripe/STPRedirectClient.m]: should we create an error here?

File: Stripe/STPRedirectClient.m [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                                                   resolvingAgainstBaseURL:NO];
+
+        NSMutableDictionary <NSString *, NSString *> *queryItems = [NSMutableDictionary new];
+        for (NSURLQueryItem *queryItem in components.queryItems) {
+            queryItems[queryItem.name] = queryItem.value;
+        }
+
+        NSString *clientSecretString = queryItems[@"client_secret"];
+//        NSString *livemodeString = queryItems[@"livemode"];
+        NSString *sourceIdString = queryItems[@"source"];
```
Comment (bg-stripe) [inline Stripe/STPRedirectClient.m]: it'd be nice to extract this query item parsing stuff to a category (e.g. on `NSURL`) and test it

File: Stripe/STPURLCallbackHandler.m [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        self.callbacks = [NSArray new];
+    }
+    return self;
+}
+
+- (BOOL)urlComponent:(NSURLComponents *)lhsComponents matchesURL:(NSURLComponents *)rhsComponents {
```
Comment (bg-stripe) [inline Stripe/STPURLCallbackHandler.m]: nit: could put this in an `NSURLComponents` category & test it

File: Stripe/PublicHeaders/STPAPIClient.h (hunk 1 above)
Comment (bdorfman-stripe) [inline Stripe/PublicHeaders/STPAPIClient.h]: Sure thing.

File: Stripe/STPRedirectClient.h (hunk 2 above)
Comment (bdorfman-stripe) [inline Stripe/STPRedirectClient.h]: Oh hmm, this is how the new built-in Xcode documenter does it. Unless there's a setting somewhere and my Xcode defaults are different.

File: Stripe/STPRedirectClient.m (hunk 3 above)
Comment (bdorfman-stripe) [inline Stripe/STPRedirectClient.m]: Yeah I am not 100% sure on all the errors yet. Also debating whether there should just be a "succeeded" BOOL or something, but not sure how to detect easily whether a redirect "completed" or if we're still waiting for confirmation. Will go through and fill in errors later once the shape of the thing is more fleshed out.

File: Stripe/STPRedirectClient.m (hunk 4 above)
Comment (bdorfman-stripe) [inline Stripe/STPRedirectClient.m]: What part specifically, converting queryItems to a dictionary?

File: Stripe/STPURLCallbackHandler.m (hunk 5 above)
Comment (bdorfman-stripe) [inline Stripe/STPURLCallbackHandler.m]: 👍

File: Stripe/STPRedirectClient.m (hunk 4 above)
Comment (bg-stripe) [inline Stripe/STPRedirectClient.m]: yeah, I was imagining something like `- (NSDictionary *)queryItems`

File: Stripe/STPRedirectClient.h (hunk 2 above)
Comment (bg-stripe) [inline Stripe/STPRedirectClient.h]: ah, I still haven't tried the built-in documenter, I just copy-paste another docstring :/

Comment (bdorfman-stripe) [inline Stripe/STPRedirectClient.h]: Don't see a way to change the Xcode behavior, and I would rather standardize around the way the official tools do it if thats ok.

File: Stripe/STPRedirectClient.m (hunk 4 above)
Comment (bdorfman-stripe) [inline Stripe/STPRedirectClient.m]: 👍

File: Stripe/STPRedirectClient.h (hunk 2 above)
Comment (bg-stripe) [inline Stripe/STPRedirectClient.h]: cool, that's fine with me

**Comment by bdorfman-stripe** [comment]
Added to category and added tests

File: Stripe/NSURLComponents+Stripe.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface NSURLComponents (Stripe)
+
+
+/**
+ Returns or sets self.queryItems as a dictionary where all the keys are the item
+ names and the values are the values. When reading, if there are duplicate
```
Comment (bg-stripe) [inline Stripe/NSURLComponents+Stripe.h]: Should this be `When writing`?

Comment (bdorfman-stripe) [inline Stripe/NSURLComponents+Stripe.h]: Nope, reading is correct. When writing you can only have one of each key, but when reading it is converting from an array which allows dupes and has significant ordering to a dictionary which does not, so.

Comment (bg-stripe) [inline Stripe/NSURLComponents+Stripe.h]: oh right, got it – missed that it was converting from a queryItems array

**Pull request merged**